Classes/MapEditor/Stats/MapStats.cs

```diff
@@ -0,0 +1,125 @@
+            this.songDuration = songDuration;
+        }
+
+        public void Recalculate(SortedSet<Note> notes, SortedSet<Note> selectedNotes) {
+            RecalculateNotes(notes, selectedNotes);
+            RecalculateColumnVariety(notes);
+            RecalculateNPS(notes);
+        }
+
+        public void RecalculateNotes(SortedSet<Note> notes, SortedSet<Note> selectedNotes) {
+            this.allNotes = notes.Count;
+            this.selectedNotes = selectedNotes.Count;
+            this.singleNotes = CalculateRowNotes(notes, 1);
+            this.doubleNotes = CalculateRowNotes(notes, 2);
+        }
+
+        private int CalculateRowNotes(SortedSet<Note> notes, int rowCount) {
+            int rowNoteCount = 0;
+            var noteIt = notes.GetEnumerator();
+            Note? lineStartNote = null;
+            var lineNoteCount = 0;
+
+
+            while (noteIt.MoveNext()) {
+                if (lineStartNote == null || !Helper.DoubleApproxEqual(noteIt.Current.beat, lineStartNote.beat)) {
+                    if (lineNoteCount == rowCount) {
+                        rowNoteCount++;
+                    }
+                    lineStartNote = noteIt.Current;
+                    lineNoteCount = 1;
+                } else {
+                    lineNoteCount++;
+                }
+            }
+            // last row to consider after all the notes were processed
+            if (lineNoteCount == rowCount) {
+                rowNoteCount++;
+            }
+
+            return rowNoteCount;
+        }
+
+        public void RecalculateColumnVariety(SortedSet<Note> notes) {
+            for (int i = 0; i < 4; i++) {
+                this.columnCounts[i] = notes.Where(note => note.col == i).Count();
+                this.columnPercentages[i] = double.Round(100 * (double)this.columnCounts[i] / Math.Max(this.allNotes, 1), 1);
+            }
+        }
+
+        public void RecalculateNPS(SortedSet<Note> notes) {
+            this.npsSong = double.Round(this.allNotes / this.songDuration, 2);
+            this.npsMapped = CalculateNPSMapped(notes);
+            this.nps16Beat = CalculateNPSPeakBeat(notes, 16);
+            this.nps8Beat = CalculateNPSPeakBeat(notes, 8);
+            this.nps4Beat = CalculateNPSPeakBeat(notes, 4);
+        }
+
+        private double CalculateNPSPeakBeat(SortedSet<Note> notes, double beats) {
+            var peakNPB = 0.0;
+            var startIt = notes.GetEnumerator();
+            startIt.MoveNext();
+            var startIndex = 0;
+            var noteIt = notes.GetEnumerator();
+            var noteIndex = 0;
+
+            while (noteIt.MoveNext()) {
+                while (Helper.DoubleApproxGreater(noteIt.Current.beat - startIt.Current.beat, beats)) {
+                    startIt.MoveNext();
+                    startIndex++;
+                }
+                peakNPB = Math.Max(peakNPB, (noteIndex - startIndex + 1) / beats);
+                noteIndex++;
+            }
```

Comment:
i would prefer to do this immutable:
```csharp
var peakNPB = notes
    .Select((note, noteIndex) => notes
        .TakeWhile((_, startIndex) => Helper.DoubleApproxGreater(note.beat - notes.ElementAt(startIndex).beat, beats))
        .Count())
    .Select((noteCount, noteIndex) => (noteIndex - noteCount + 1) / beats)
    .DefaultIfEmpty(0)
    .Max();
```

Reply:
This certainly works, but is **very** noticably slower with around 1000 notes. The `foreach` approach has linear time complexity, while the LINQ version is quadratic.

Reply:
Nevermind, I figured out a faster way to do it with `SortedSet` :D